odor/src/test/scala/DockerTests.scala

```diff
@@ -0,0 +1,129 @@
+          assert(res == 1337)
+        }
+      }
+    }
+  }
+
+  "PostgresClient" should "correctly set the transaction isolation level" in async {
+    await(
+      Future.sequence(
+        for (
+          isolationLevel <-
+            Vector(IsolationLevel.ReadCommitted, IsolationLevel.RepeatableRead, IsolationLevel.Serializable)
+        ) yield {
+          pool.useConnection(isolationLevel = isolationLevel) { pgClient =>
+            assertCurrentIsolationLevel(isolationLevel, pgClient)
+          }
+        },
+      ),
+    ): Unit
+
+    succeed
+  }
+
+  "PostgresConnectionPool" should "set a default isolation level" in async {
+    val defaultLevel = IsolationLevel.RepeatableRead
+
+    val pool: PostgresConnectionPool[IsolationLevel.RepeatableRead] = PostgresConnectionPool(
+      PG_CONNECTION_STRING,
+      maxConnections = 2,
+      defaultIsolationLevel = defaultLevel,
+    )
+
+    await(pool.useConnection() { pgClient =>
+      assertCurrentIsolationLevel(defaultLevel, pgClient)
+    })
+
+    val overrideLevel = IsolationLevel.Serializable
+
+    await(pool.useConnection(isolationLevel = overrideLevel) { pgClient =>
+      assertCurrentIsolationLevel(overrideLevel, pgClient)
+    })
+
+    import org.scalatest.matchers.should.Matchers._
+
+    @unused
+    def serializable(
+      @unused p: PostgresClient { type TransactionIsolationLevel <: IsolationLevel.Serializable },
+    ): Future[Unit] = Future.unit
+
+    """
+    pool.useConnection() { pgClient =>
+      serializable(pgClient)
+    }
+    """ shouldNot typeCheck
+
+    """
+    pool.useConnection(isolationLevel = overrideLevel) { pgClient =>
+      serializable(pgClient)
+    }
+    """ should compile
+  }
+
+  "PostgresConnectionPool" should "propagate the default isolation level at compile time" in async {
+    import org.scalatest.matchers.should.Matchers._
+
+    // helper function that should only compile if the given `PostgresClient` has at least isolation level
+    // `RepeatableRead`.
+    @unused
+    def repeatableRead(
+      @unused p: PostgresClient { type TransactionIsolationLevel <: IsolationLevel.RepeatableRead },
+    ): Future[Unit] = Future.unit
+
+    // connection pool that has a default isolation level of `RepeatableRead`
+    @unused
+    val pool: PostgresConnectionPool[IsolationLevel.RepeatableRead] = PostgresConnectionPool(
+      PG_CONNECTION_STRING,
+      maxConnections = 2,
+      defaultIsolationLevel = IsolationLevel.RepeatableRead,
+    )
+
+    // TODO: this should work/compile, but it currently does not. it looks like the compiler
+    // can't figure out that the type we put together in `pool.useConnection()` contains an isolation
+    // level of `RepeatableRead` from the default isolation level of the pool. See `PostgresConnectionPool.useConnection()`.
+    """
+    pool.useConnection() { pgClient =>
+      repeatableRead(pgClient)
+    }
+    """ should compile
+  }
+}
```

Comment:
@cornerman this right here should work, but currently doesn't. `PostgresConnectionPool.useConnection()` contains the relevant implementation parts (I think). Thank you for taking a look! :)

Reply:
The type construction in the `useConnection` method looks like the compiler cannot proof that the (default) isolationLevel that is passed in, is actually the same as the one from the `PostgresConnectionPool`. 

I would try using generics on the `useConnection` there: Maybe something like this? (Though, I have not tried it yet)

```
  @nowarn("msg=unused value")
  def useConnection[R](
    code: PostgresClient {
      type TransactionIsolationLevel <: I;
    } => Future[R],
  ): Future[R] = useConnection[I, R](defaultIsolationLevel)(code)
  
  @nowarn("msg=unused value")
  def useConnection[Iso <: I, R](
    isolationLevel: Iso,
  )(
    code: PostgresClient {
      type TransactionIsolationLevel <: Iso;
    } => Future[R],
  ): Future[R] = ???
```

This way, the compiler can hopefully track the `Iso` generic on the call-side, and the `I` type is also transparently passed from the `PostgresConnectionPool` to the `PostgresClient`.

Reply:
Awesome, trying it out right now, and so far it seems to work! Only part is that this is not 100% extensible for more (default) arguments, but I guess it's fine for now! :) Thanks a lot! :D